### ISSUE TITLE
Java Execution compilation gradually gets slower 

### ISSUE THREAD

**Issue by simmel** [elastic/logstash#9255]
- Version: 6.2.1
- Operating System: Debian stretch
- JVM version: 1.8.0_152
- Config File (if you have sensitive info, please remove it): config.conf in the tar-file.
- Sample Data: N/A
- Steps to Reproduce: `bin/logstash --experimental-java-execution -f config.conf --log.level=debug`

For jstacks and config.conf see https://stockholmuniversity.box.com/shared/static/uqokj5f72l8xbyhlhbe7imztn4fxdrqa.tar

**Comment by simmel** [comment]
While I've mentioned it before elsewhere but I thought I'd mention it here too. It seems that the compilation goes slower and slower and/or longer between compilations.

Our full ruleset hadn't compiled after 50 minutes.

**Comment by original-brownbear** [comment]
@simmel thanks for the config here. Will work on this today. Looks like there is a principle bug in how we compile the pipeline to Java here.

**Comment by original-brownbear** [comment]
@simmel I can't reproduce the problem here. Running your config works correctly for me and starts in less than a minute:

```sh
HOSTNAME=localhost bin/logstash -w 100 -b 128  -f ~/tmp/config.conf
Sending Logstash's logs to /Users/brownbear/src/logstash/logs which is now configured via log4j2.properties
[2018-03-21T09:49:39,308][INFO ][logstash.modules.scaffold] Initializing module {:module_name=>"netflow", :directory=>"/Users/brownbear/src/logstash/modules/netflow/configuration"}
[2018-03-21T09:49:39,328][INFO ][logstash.modules.scaffold] Initializing module {:module_name=>"fb_apache", :directory=>"/Users/brownbear/src/logstash/modules/fb_apache/configuration"}
[2018-03-21T09:49:39,580][WARN ][logstash.config.source.multilocal] Ignoring the 'pipelines.yml' file because modules or command line options are specified
[2018-03-21T09:49:40,415][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"7.0.0-alpha1"}
[2018-03-21T09:49:40,863][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600}
[2018-03-21T09:49:49,215][INFO ][logstash.filters.geoip   ] Using geoip database {:path=>"/Users/brownbear/src/logstash/vendor/bundle/jruby/2.3.0/gems/logstash-filter-geoip-5.0.3-java/vendor/GeoLite2-City.mmdb"}
[2018-03-21T09:49:49,517][INFO ][logstash.javapipeline    ] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>100, "pipeline.batch.size"=>128, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>12800, :thread=>"#<Thread:0xa4b8c38 run>"}
[2018-03-21T09:49:49,522][WARN ][logstash.javapipeline    ] CAUTION: Recommended inflight events max exceeded! Logstash will run with up to 12800 events in memory in your current configuration. If your message sizes are large this may cause instability with the default heap size. Please consider setting a non-standard heap size, changing the batch size (currently 128), or changing the number of pipeline workers (currently 100) {:pipeline_id=>"main", :thread=>"#<Thread:0xa4b8c38 run>"}
[2018-03-21T09:49:49,600][INFO ][logstash.inputs.tcp      ] Starting tcp input listener {:address=>"localhost:2514", :ssl_enable=>"false"}
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/brownbear/src/logstash/logstash-core/lib/jars/log4j-slf4j-impl-2.6.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/brownbear/src/logstash/logstash-core/lib/jars/log4j-slf4j-impl-2.9.1.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]
[2018-03-21T09:49:49,852][INFO ][logstash.javapipeline    ] Pipeline started {"pipeline.id"=>"main"}
[2018-03-21T09:49:49,976][INFO ][logstash.agent           ] Pipelines running {:count=>1, :pipelines=>["main"]}
```

Am I missing something here in the reproducer?

**Comment by original-brownbear** [comment]
@simmel interesting ... if I use `--log.level=debug` it doesn't seem to finish (at least not within ~5min). Can you confirm whether or not it works for you without the debug logging?

Just to narrow it down, so far it seems to me that logging is the problem here.

**Comment by original-brownbear** [comment]
nevermind reproduced without logging as well, on it with a fix:)

**Comment by original-brownbear** [comment]
@andrewvc @simmel so this is what I found: 

* I can reproduce the extremely slow compilation issue
   * It is just slow though, it eventually finishes
   * Slowness is coming from 2 sources:
       * We generate the full Java source once per worker thread redundantly (there is a cache source -> class that then prevents duplicate classes from being compiled, but generating the source is done needlessly in all but one worker thread). Since there is some synchronization between threads here (mostly when accessing caches) this gets fairly slow
       * 40-60% of the runtime during compilation is taken up by calls to `IfVertex.getOutgoingBooleanEdgesByType`, so we probably have some potential for optimization here regardless of the redundant source gen.

Fixing those two now :)

**Comment by original-brownbear** [comment]
Much improved in https://github.com/elastic/logstash/pull/9268 :)

**Comment by andrewvc** [comment]
@original-brownbear I see that #9268 addresses the compilation caching issue. Any plans to address the `getBooleanEdgesByType` thing?

2mins is much improved, but I'm wondering how much more room there is for speed here? Any idea what the ratio is between TreeTop and compilation?

It might be good to add debug logs timing the different phases of compilation, like:

1. Time in TreeTop
2. Time converting imperative to graph
3. Time generating java
4. Time compiling java by janino

WDYT?

**Comment by original-brownbear** [comment]
@andrewvc I benchmarked this already :)

So rough numbers (for this example, this will vary heavily depending on the actual config) are:

Time generating Java takes about 80% of the time (after my fix, before it's 99%+).
The other operations are much faster.

There is serious room for improvement here, but I don't think there are many more quick fixes like the one I found.
One obvious way is to cut back on redundant computations of the Java source. With my fix, that brings the startup time down 50% in theory (takes ~1min with `-w 1` and ~2 min with `-w 8`). That's not so trivial to do now and will take some time to implement. I can start work on that though if we think it's worth the effort.

The `getBooleanEdgesByType` I ignored for now since it's probably just going to be a ~20-30% gain past my fix in the PR. I could add some caching for the result of that call though and it should give us a nice little boost. I'll do that in another PR?

**Comment by andrewvc** [comment]
@original-brownbear +1 to do that in another PR.

When you say time generating java is that generating the source text, or time Janino spends compiling it? It sounds like the former

**Comment by original-brownbear** [comment]
@andrewvc yea Janino itself barely takes any time.

**Comment by original-brownbear** [comment]
@simmel fixed in `master` and `6.x`, brought the time for your config down to like 2-3s :)

**Comment by andrewvc** [comment]
Awesome work here @original-brownbear !

**Comment by simmel** [comment]
On Thu, 2018-03-22 at 20:37:48 +0000, Armin Braun wrote:
>@simmel fixed in `master` and `6.x`, brought the time for your config down to like 2-3s :)

Nice! Our startup went from 2m (without java exec) to 1m (with java 
exec) with your fixes = )

**Comment by simmel** [comment]
But I get an exception and the pipeline stops @original-brownbear 

```
[2018-03-26T12:57:43,538][ERROR][org.logstash.execution.WorkerLoop] Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.
java.lang.ClassCastException: org.jruby.RubyNil cannot be cast to org.jruby.RubyArray
        at org.logstash.generated.CompiledDataset195.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset201.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset203.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset214.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset216.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset220.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset222.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset226.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset240.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset244.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset256.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset258.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset269.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset271.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset275.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset284.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset285.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset292.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset296.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset298.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset308.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset310.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset314.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset316.compute(Unknown Source) ~[?:?]
        at org.logstash.generated.CompiledDataset389.compute(Unknown Source) ~[?:?]
        at org.logstash.execution.WorkerLoop.run(WorkerLoop.java:54) [logstash-core.jar:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_151]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_151]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_151]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_151]
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:423) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:290) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:28) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:90) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145) [jruby-complete-9.1.13.0.jar:?]
        at Users.simlu.src.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$block$start_workers$2(/Users/simlu/src/logstash/logstash-core/lib/logstash/java_pipeline.rb:373) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:145) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:71) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.runtime.Block.call(Block.java:124) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.RubyProc.call(RubyProc.java:289) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.RubyProc.call(RubyProc.java:246) [jruby-complete-9.1.13.0.jar:?]
        at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:104) [jruby-complete-9.1.13.0.jar:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
Exception in thread "Ruby-0-Thread-19@[main]>worker0: :1" java.lang.IllegalStateException: java.lang.ClassCastException: org.jruby.RubyNil cannot be cast to org.jruby.RubyArray
        at org.logstash.execution.WorkerLoop.run(org/logstash/execution/WorkerLoop.java:75)
        at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:423)
        at org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:290)
        at Users.simlu.src.logstash.logstash_minus_core.lib.logstash.java_pipeline.block in start_workers(/Users/simlu/src/logstash/logstash-core/lib/logstash/java_pipeline.rb:373)
        at org.jruby.RubyProc.call(org/jruby/RubyProc.java:289)
        at org.jruby.RubyProc.call(org/jruby/RubyProc.java:246)
        at java.lang.Thread.run(java/lang/Thread.java:748)
Caused by: java.lang.ClassCastException: org.jruby.RubyNil cannot be cast to org.jruby.RubyArray
        at org.logstash.generated.CompiledDataset195.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset201.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset203.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset214.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset216.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset220.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset222.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset226.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset240.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset244.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset256.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset258.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset269.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset271.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset275.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset284.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset285.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset292.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset296.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset298.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset308.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset310.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset314.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset316.compute(Unknown Source)
        at org.logstash.generated.CompiledDataset389.compute(Unknown Source)
        at org.logstash.execution.WorkerLoop.run(WorkerLoop.java:54)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:423)
        at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:290)
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:28)
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:90)
        at org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:145)
        at Users.simlu.src.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$block$start_workers$2(/Users/simlu/src/logstash/logstash-core/lib/logstash/java_pipeline.rb:373)
        at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:145)
        at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:71)
        at org.jruby.runtime.Block.call(Block.java:124)
        at org.jruby.RubyProc.call(RubyProc.java:289)
        at org.jruby.RubyProc.call(RubyProc.java:246)
        at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:104)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by original-brownbear** [comment]
@simmel but it runs without any exception with the Ruby exec?

**Comment by simmel** [comment]
Better/appropriate names would be desirable for the compiled rules to be able to further narrow down which rule is the problem here = )

**Comment by simmel** [comment]
It starts up fine but then it throws that exception when it hits some compiled rule I guess?

**Comment by original-brownbear** [comment]
@simmel yea, better logging is on my list here :) That said, we shouldn't have an error in Java execution that we don't have in Ruby execution if that's the case here?

**Comment by simmel** [comment]
The exception comes even though I don't put anything in the console/stdin.

That's is the case. I get no exception with ruby exec.

**Comment by original-brownbear** [comment]
@simmel oh ok, thanks for reporting that then. Will take a look, seems we have a remaining issue here then.

**Comment by simmel** [comment]
And both on Linux and OS X if that's any comfort = D

Feel free to hit me up on IRC if you want me to try something or have quicker response.

**Comment by original-brownbear** [comment]
@simmel thanks, got it reproduced already :)

**Comment by simmel** [comment]
Woo, let me know when you have commited it.

**Comment by original-brownbear** [comment]
@simmel @andrewvc tracked this down :) The problem is that I forgot to add a check to the flush call in the pipeline compiler.

In the Ruby version code like this is generated:

```rb
              events = @generated_objects[:filter_dlq_commit_2].flush(options)
              return if events.nil? || events.empty?
```

that breaks out if a `flush` call returns `nil`. The Java execution mistakenly assumes to always get an `[]` here.
Will add that check :)

**Comment by original-brownbear** [comment]
@simmel fix for this was merged to `master` and `6.x` :)

**Comment by simmel** [comment]
On Mon, 2018-03-26 at 07:48:13 -0700, Armin Braun wrote:
>@simmel fix for this was merged to `master` and `6.x` :)

Woo, works perfectly! Thanks!

**Comment by simmel** [comment]
Uh.. something is wrong.. I'll reconnect when I know more.

**Comment by simmel** [comment]
I tried in with our full ruleset and everything exploded. ES started whining about too many fields (>1000) and Logstash just spewed log.

I tried the `config.conf` that I have above and LS uses 100% CPU on all workers.
A threaddump on the workers says everyone of them is doing:
```
"Ruby-0-Thread-19@[main]>worker0: :1" #37 daemon prio=5 os_prio=31 tid=0x00007fe450ed3000 nid=0x7b03 runnable [0x000070000e23e000]
   java.lang.Thread.State: RUNNABLE
        at org.jruby.runtime.Helpers.fillNil(Helpers.java:2051)
        at org.jruby.RubyArray.rb_clear(RubyArray.java:2114)
        at org.jruby.RubyArray.clear(RubyArray.java:4861)
        at org.logstash.generated.CompiledDataset11.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset14.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset17.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset20.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset26.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset29.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset32.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset35.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset38.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset41.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset44.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset50.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset56.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset65.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset74.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset77.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset86.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset98.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset110.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset113.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset122.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset125.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset134.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset146.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset164.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset167.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset170.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset187.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset191.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset206.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset209.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset212.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset233.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset248.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset260.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset263.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset266.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset278.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset290.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset293.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset296.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset305.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset308.clear(Unknown Source)
        at org.logstash.generated.CompiledDataset311.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset332.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset332.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset332.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset332.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset332.clear(Unknown Source)
        at org.logstash.config.ir.compiler.DatasetCompiler$Complement.clear(DatasetCompiler.java:423)
        at org.logstash.generated.CompiledDataset389.compute(Unknown Source)
        at org.logstash.execution.WorkerLoop.run(WorkerLoop.java:54)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:438)
        at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:302)
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:36)
        at java.lang.invoke.LambdaForm$DMH/409007038.invokeVirtual_L5_L(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$BMH/728370857.reinvoke(LambdaForm$BMH)
        at java.lang.invoke.LambdaForm$reinvoker/832743930.dontInline(LambdaForm$reinvoker)
        at java.lang.invoke.LambdaForm$MH/834153999.guard(LambdaForm$MH)
        at java.lang.invoke.LambdaForm$reinvoker/832743930.dontInline(LambdaForm$reinvoker)
        at java.lang.invoke.LambdaForm$MH/834153999.guard(LambdaForm$MH)
        at java.lang.invoke.LambdaForm$MH/977195776.linkToCallSite(LambdaForm$MH)
        at Users.simlu.src.logstash.logstash_minus_core.lib.logstash.java_pipeline.RUBY$block$start_workers$2(/Users/simlu/src/logstash/logstash-core/lib/logstash/java_pipeline.rb:373)
        at java.lang.invoke.LambdaForm$DMH/1914301543.invokeStatic_L8_L(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/1106933404.invokeExact_MT(LambdaForm$MH)
        at org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:145)
        at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:71)
        at org.jruby.runtime.Block.call(Block.java:124)
        at org.jruby.RubyProc.call(RubyProc.java:289)
        at org.jruby.RubyProc.call(RubyProc.java:246)
        at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:104)
        at java.lang.Thread.run(Thread.java:748)
```

Also found another Java exec json filter related issue, should I file a seperate issue for this or should I continue in this one?

**Comment by original-brownbear** [comment]
@simmel new one would be nice, please ping me in there and I'll take a look shortly :)

**Comment by simmel** [comment]
One for the `fillNill`-issue and one for the json filter one or?

**Comment by original-brownbear** [comment]
@simmel just one for the json-filter is fine, looking into the fill-nil thing here :)

**Comment by simmel** [comment]
Cool

**Comment by original-brownbear** [comment]
@simmel redundant `clear` calls also fixed in #9294 :)
